Review commit

yuliagurevich opened the pull request.

yuliagurevich committed:
Review commit

gennady-bars commented on src/components/app-header/app-header.tsx:
```diff
@@ -1,4 +1,15 @@
+import { useSelector } from '@store';
+import { useLocation } from 'react-router-dom';
+import { userSelectors } from '@slices';
 
```
**Можно лучше**


Если интересно, то можно создать **универсальную** функцию запроса `request` , которая будет содержать в себе уже **базовый урл**, проверку на **ok** и проверку на `success`, чтобы не дублировать их в каждом запросе.

```js

// 1 раз объявляем базовый урл
export const BASE_URL = "https://norma.nomoreparties.space/api/";

// создаем функцию проверки ответа на `ok`
const checkResponse = (res) => {
  if (res.ok) {
    return res.json();
  }
  // не забываем выкидывать ошибку, чтобы она попала в `catch`
  return Promise.reject(`Ошибка ${res.status}`);
};

// создаем функцию проверки на `success`
const checkSuccess = (res) => {
  if (res && res.success) {
    return res;
  }
  // не забываем выкидывать ошибку, чтобы она попала в `catch`
  return Promise.reject(`Ответ не success: ${res}`);
};

// создаем универсальную фукнцию запроса с проверкой ответа и `success`
// В вызов приходит `endpoint`(часть урла, которая идет после базового) и опции
const request = (endpoint, options) => {
  // а также в ней базовый урл сразу прописывается, чтобы не дублировать в каждом запросе
  return fetch(`${BASE_URL}${endpoint}`, options)
    .then(checkResponse)
    .then(checkSuccess);
};

```


И теперь не нужно будет дублировать все это в каждом запросе. Нужно просто заменить в запросах все `fetch` на `request`, удалив лишние проверки и базовый урл. Все остальное останется без изменений

Код станет намного чище. Будет мелкий аналог `axios`

Пример запроса ингредиентов

```js
const getIngredients =  () => request("ingredients");

```

В `get`-запросах даже не нужно указывать 2й параметр. Он получится `undefined`, что без проблем для `fetch` (который внутри).

yuliagurevich replied:
Интересно. Спасибо за рекомендацию. Обязательно попробую реализовать. Понимаю, что плохо усвоила работу промисов. Выделю на это время вне проектной работы.

gennady-bars commented on src/pages/profile-orders/profile-orders.tsx:
```diff
@@ -1,10 +1,19 @@
 import { TOrder } from '@utils-types';
-import { FC } from 'react';
+import { FC, useEffect } from 'react';
+import { useSelector } from 'react-redux';
```
`useSelector` нужно импортировать из файла `store.ts` (он там типизированный уже)

gennady-bars commented on src/services/slices/burger.ts:
```diff
@@ -0,0 +1,74 @@
+          state.constructorItems.ingredients.push(action.payload);
+        }
+      },
+      prepare: (ingredient: TIngredient) => ({
+        payload: { ...ingredient, id: uuid() }
```
Отлично, что в `prepare` создаете `id`

gennady-bars commented on src/pages/login/login.tsx:
```diff
@@ -1,17 +1,37 @@
+import { useLocation } from 'react-router-dom';
 
 export const Login: FC = () => {
   const [email, setEmail] = useState('');
   const [password, setPassword] = useState('');
```
**Можно лучше**

Можно сделать универсальный **кастомный** хук для контроля любого количества инпутов в любых формах:

```js
export function useForm(inputValues={}) {
  const [values, setValues] = useState(inputValues);

  const handleChange = (event) => {
    const {value, name} = event.target;
    setValues({...values, [name]: value});
  };
  return {values, handleChange, setValues};
}

```

Этот код помещают в отдельный файл `useForm.js` в папке `hooks` и импортируют функцию туда, где нужно контролировать инпуты

И Вам не нужно будет теперь вручную создавать функции обработки инпутов и т д. Все будет в одной строчке кода:

```jsx
  const {values, handleChange, setValues} = useForm({});

```

yuliagurevich replied:
Согласна, так лучше. Но оставлю не свою часть кода без изменений. Возьму себе на заметку в своих проектах. Спасибо.

gennady-bars commented on src/components/protected-route/protected-route.tsx:
```diff
@@ -0,0 +1,51 @@
+      />
+    );
+  }
+
+  // The user is authenticated and is redirected to the page he came from or to the root route
+  if (onlyUnAuth && user) {
+    const from = location.state?.from || { pathname: '/' };
+    const background = location.state?.from?.background || null;
+    return <Navigate to={from} state={{ background }} replace />;
```
Хорошо сделали `ProtectedRoute`

gennady-bars commented on src/components/app/app.tsx:
```diff
@@ -1,14 +1,144 @@
+  const navigate = useNavigate();
+  const location = useLocation();
+
+  const backgroundPath = location.state?.background;
```
Отлично сделана логика отображения попапа и страницы  под ним

gennady-bars commented on src/services/slices/user.ts:
```diff
@@ -0,0 +1,144 @@
+  isAuthChecked: false
+};
+
+const registerUser = createAsyncThunk(
+  'user/register',
+  async (data: TRegisterData) => registerUserApi(data)
+);
+
+const loginUser = createAsyncThunk('user/login', async (data: TLoginData) =>
+  loginUserApi(data)
+);
+
+const getUser = createAsyncThunk('user/get', async () => getUserApi());
+
+const updateUser = createAsyncThunk(
+  'user/update',
+  async (data: Partial<TRegisterData>) => updateUserApi(data)
+);
+
+const logoutUser = createAsyncThunk('user/logout', async () => logoutApi());
+
```
**Можно лучше**

* Если внутри **безымянной** (стрелочной) функции вызывается одна функция с **точно такими же** аргументами, то эта безымянная функция не нужна. Это лишняя обертка вокруг самой функции, которую можно использовать просто сразу в коде



```ts

const registerUser = createAsyncThunk('user/register', registerUserApi);

const loginUser = createAsyncThunk('user/login', loginUserApi);

const getUser = createAsyncThunk('user/get', getUserApi);

const updateUser = createAsyncThunk('user/update', updateUserApi);

const logoutUser = createAsyncThunk('user/logout', logoutApi);
```

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная:




* Отлично, что в `prepare` создаете `id`
* Хорошая логика попапов
* Отлично сделана логика отображения попапа и страницы  под ним
* Хорошо сделали `ProtectedRoute`


но есть некоторые недочеты:



* `useSelector` нужно импортировать из файла `store.ts` (он там типизированный уже)
* В попапе ингредиента должна быть надпись https://disk.yandex.ru/i/ZZ4aBc0QyR3GgA 


**Можно лучше**

* при клике на **логотип** обычно идет переход на **главную** страницу
* Если внутри **безымянной** (стрелочной) функции вызывается одна функция с **точно такими же** аргументами, то эта безымянная функция не нужна. Это лишняя обертка вокруг самой функции, которую можно использовать просто сразу в коде

yuliagurevich committed:
- Added auto-open feature to webpack server configuration.
- Added a header for ingredient details.
- Implemented navigation to the homepage on logo click.
- Fixed a bug where constructor data wasn't resetting.
- Removed unnecessary use of arrow functions in API calls.
- Typed actions in Redux slices.

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
